### ISSUE TITLE
Checked out master on OSX 10.11.6 with Rust 1.11.0 unable to compile cube or triangle example

### ISSUE THREAD

**Issue by jamiltron** [gfx-rs/gfx#1048]
Hello,

I just cloned gfx-rs on my OSX 10.11.6 machine, tried running the examples, and ran into a number of errors.

Here's the steps I used to arrive at the errors:
```
git clone https://github.com/gfx-rs/gfx.git
cd gfx
cargo run --example cube
````

The errors I'm seeing look related to the metal backend:
```
src/backend/metal/src/command.rs:25:5: 25:27 error: unresolved import `gfx_core::shade::Stage`. Maybe a missing `extern crate gfx_core`? [E0432]
src/backend/metal/src/command.rs:25 use gfx_core::shade::Stage;
                                        ^~~~~~~~~~~~~~~~~~~~~~
src/backend/metal/src/command.rs:25:5: 25:27 help: run `rustc --explain E0432` to see a detailed explanation
src/backend/metal/src/map.rs:443:9: 443:38 error: unresolved import `ore::shade::IsMultiSample::*`. Maybe a missing `extern crate ore`? [E0432]
src/backend/metal/src/map.rs:443     use ore::shade::IsMultiSample::*;
                                         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
src/backend/metal/src/map.rs:443:9: 443:38 help: run `rustc --explain E0432` to see a detailed explanation
src/backend/metal/src/factory.rs:40:6: 40:24 error: `core::mapping::Raw` is not a trait [E0404]
src/backend/metal/src/factory.rs:40 impl core::mapping::Raw for RawMapping {
                                         ^~~~~~~~~~~~~~~~~~
src/backend/metal/src/factory.rs:40:6: 40:24 help: run `rustc --explain E0404` to see a detailed explanation
src/backend/metal/src/factory.rs:77:44: 77:63 error: type name `factory::BufferInfo` is undefined or not in scope [E0412]
src/backend/metal/src/factory.rs:77     fn create_buffer_internal(&self, info: factory::BufferInfo, raw_data: Option<*const c_void>)
                                                                               ^~~~~~~~~~~~~~~~~~~
src/backend/metal/src/factory.rs:77:44: 77:63 help: run `rustc --explain E0412` to see a detailed explanation
src/backend/metal/src/factory.rs:77:44: 77:63 help: no candidates by the name of `BufferInfo` found in your project; maybe you misspelled the name or forgot to import an external crate?
src/backend/metal/src/factory.rs:78:53: 78:73 error: type name `factory::BufferError` is undefined or not in scope [E0412]
src/backend/metal/src/factory.rs:78             -> Result<handle::RawBuffer<Resources>, factory::BufferError> {
                                                                                        ^~~~~~~~~~~~~~~~~~~~
src/backend/metal/src/factory.rs:78:53: 78:73 help: run `rustc --explain E0412` to see a detailed explanation
src/backend/metal/src/factory.rs:78:53: 78:73 help: no candidates by the name of `BufferError` found in your project; maybe you misspelled the name or forgot to import an external crate?
src/backend/metal/src/factory.rs:83:31: 83:53 error: unresolved name `factory::RENDER_TARGET` [E0425]
src/backend/metal/src/factory.rs:83         if info.bind.contains(factory::RENDER_TARGET) | info.bind.contains(factory::DEPTH_STENCIL) {
                                                                  ^~~~~~~~~~~~~~~~~~~~~~
src/backend/metal/src/factory.rs:83:31: 83:53 help: run `rustc --explain E0425` to see a detailed explanation
src/backend/metal/src/factory.rs:83:76: 83:98 error: unresolved name `factory::DEPTH_STENCIL` [E0425]
src/backend/metal/src/factory.rs:83         if info.bind.contains(factory::RENDER_TARGET) | info.bind.contains(factory::DEPTH_STENCIL) {
                                                                                                               ^~~~~~~~~~~~~~~~~~~~~~
src/backend/metal/src/factory.rs:83:76: 83:98 help: run `rustc --explain E0425` to see a detailed explanation
src/backend/metal/src/factory.rs:84:24: 84:61 error: failed to resolve. Could not find `BufferError` in `core::factory` [E0433]
src/backend/metal/src/factory.rs:84             return Err(factory::BufferError::UnsupportedBind(info.bind))
                                                           ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
src/backend/metal/src/factory.rs:84:24: 84:61 help: run `rustc --explain E0433` to see a detailed explanation
src/backend/metal/src/factory.rs:141:5: 141:30 error: type `Mapper` is not a member of trait `core::Factory` [E0437]
src/backend/metal/src/factory.rs:141     type Mapper = RawMapping;
                                         ^~~~~~~~~~~~~~~~~~~~~~~~~
src/backend/metal/src/factory.rs:141:5: 141:30 help: run `rustc --explain E0437` to see a detailed explanation
src/backend/metal/src/factory.rs:147:43: 147:62 error: type name `factory::BufferInfo` is undefined or not in scope [E0412]
src/backend/metal/src/factory.rs:147     fn create_buffer_raw(&mut self, info: factory::BufferInfo) -> Result<handle::RawBuffer<Resources>, factory::BufferError> {
                                                                               ^~~~~~~~~~~~~~~~~~~
src/backend/metal/src/factory.rs:147:43: 147:62 help: run `rustc --explain E0412` to see a detailed explanation
src/backend/metal/src/factory.rs:147:43: 147:62 help: no candidates by the name of `BufferInfo` found in your project; maybe you misspelled the name or forgot to import an external crate?
src/backend/metal/src/factory.rs:147:104: 147:124 error: type name `factory::BufferError` is undefined or not in scope [E0412]
src/backend/metal/src/factory.rs:147     fn create_buffer_raw(&mut self, info: factory::BufferInfo) -> Result<handle::RawBuffer<Resources>, factory::BufferError> {
                                                                                                                                            ^~~~~~~~~~~~~~~~~~~~
src/backend/metal/src/factory.rs:147:104: 147:124 help: run `rustc --explain E0412` to see a detailed explanation
src/backend/metal/src/factory.rs:147:104: 147:124 help: no candidates by the name of `BufferError` found in your project; maybe you misspelled the name or forgot to import an external crate?
src/backend/metal/src/factory.rs:151:5: 161:6 error: method `create_buffer_const_raw` is not a member of trait `core::Factory` [E0407]
src/backend/metal/src/factory.rs:151     fn create_buffer_const_raw(&mut self, data: &[u8], stride: usize, role: factory::BufferRole, bind: factory::Bind)
                                         ^
src/backend/metal/src/factory.rs:151:5: 161:6 help: run `rustc --explain E0407` to see a detailed explanation
src/backend/metal/src/factory.rs:151:77: 151:96 error: type name `factory::BufferRole` is undefined or not in scope [E0412]
src/backend/metal/src/factory.rs:151     fn create_buffer_const_raw(&mut self, data: &[u8], stride: usize, role: factory::BufferRole, bind: factory::Bind)
                                                                                                                 ^~~~~~~~~~~~~~~~~~~
src/backend/metal/src/factory.rs:151:77: 151:96 help: run `rustc --explain E0412` to see a detailed explanation
src/backend/metal/src/factory.rs:151:77: 151:96 help: no candidates by the name of `BufferRole` found in your project; maybe you misspelled the name or forgot to import an external crate?
src/backend/metal/src/factory.rs:151:104: 151:117 error: type name `factory::Bind` is undefined or not in scope [E0412]
src/backend/metal/src/factory.rs:151     fn create_buffer_const_raw(&mut self, data: &[u8], stride: usize, role: factory::BufferRole, bind: factory::Bind)
                                                                                                                                            ^~~~~~~~~~~~~
src/backend/metal/src/factory.rs:151:104: 151:117 help: run `rustc --explain E0412` to see a detailed explanation
src/backend/metal/src/factory.rs:151:104: 151:117 help: you can import it into scope: `use core::memory::Bind;`.
src/backend/metal/src/factory.rs:152:73: 152:93 error: type name `factory::BufferError` is undefined or not in scope [E0412]
src/backend/metal/src/factory.rs:152                                 -> Result<handle::RawBuffer<Resources>, factory::BufferError> {
                                                                                                             ^~~~~~~~~~~~~~~~~~~~
src/backend/metal/src/factory.rs:152:73: 152:93 help: run `rustc --explain E0412` to see a detailed explanation
src/backend/metal/src/factory.rs:152:73: 152:93 help: no candidates by the name of `BufferError` found in your project; maybe you misspelled the name or forgot to import an external crate?
src/backend/metal/src/factory.rs:153:20: 153:39 error: `factory::BufferInfo` does not name a structure [E0422]
src/backend/metal/src/factory.rs:153         let info = factory::BufferInfo {
                                                        ^~~~~~~~~~~~~~~~~~~
src/backend/metal/src/factory.rs:153:20: 153:39 help: run `rustc --explain E0422` to see a detailed explanation
src/backend/metal/src/factory.rs:155:20: 155:41 error: failed to resolve. Could not find `Usage` in `core::factory` [E0433]
src/backend/metal/src/factory.rs:155             usage: factory::Usage::Const,
                                                        ^~~~~~~~~~~~~~~~~~~~~
src/backend/metal/src/factory.rs:155:20: 155:41 help: run `rustc --explain E0433` to see a detailed explanation
src/backend/metal/src/factory.rs:355:44: 355:69 error: type name `core::texture::Descriptor` is undefined or not in scope [E0412]
src/backend/metal/src/factory.rs:355     fn create_texture_raw(&mut self, desc: core::texture::Descriptor, hint: Option<core::format::ChannelType>,
                                                                                ^~~~~~~~~~~~~~~~~~~~~~~~~
src/backend/metal/src/factory.rs:355:44: 355:69 help: run `rustc --explain E0412` to see a detailed explanation
src/backend/metal/src/factory.rs:355:44: 355:69 help: you can import it into scope: `use core::pso::Descriptor;`.
src/backend/metal/src/factory.rs:356:96: 356:116 error: type name `core::texture::Error` is undefined or not in scope [E0412]
src/backend/metal/src/factory.rs:356                           data_opt: Option<&[&[u8]]>) -> Result<handle::RawTexture<Resources>, core::texture::Error> {
                                                                                                                                    ^~~~~~~~~~~~~~~~~~~~
src/backend/metal/src/factory.rs:356:96: 356:116 help: run `rustc --explain E0412` to see a detailed explanation
src/backend/metal/src/factory.rs:356:96: 356:116 help: you can import several candidates into scope (`use ...;`):
src/backend/metal/src/factory.rs:356:96: 356:116 help:   `core::mapping::Error`
src/backend/metal/src/factory.rs:356:96: 356:116 help:   `std::error::Error`
src/backend/metal/src/factory.rs:356:96: 356:116 help:   `std::fmt::Error`
src/backend/metal/src/factory.rs:356:96: 356:116 help:   `std::io::Error`
src/backend/metal/src/factory.rs:671:80: 671:98 error: type name `factory::MapAccess` is undefined or not in scope [E0412]
src/backend/metal/src/factory.rs:671     fn map_buffer_raw(&mut self, _buf: &handle::RawBuffer<Resources>, _access: factory::MapAccess) -> RawMapping {
                                                                                                                    ^~~~~~~~~~~~~~~~~~
src/backend/metal/src/factory.rs:671:80: 671:98 help: run `rustc --explain E0412` to see a detailed explanation
src/backend/metal/src/factory.rs:671:80: 671:98 help: no candidates by the name of `MapAccess` found in your project; maybe you misspelled the name or forgot to import an external crate?
src/backend/metal/src/factory.rs:675:5: 677:6 error: method `unmap_buffer_raw` is not a member of trait `core::Factory` [E0407]
src/backend/metal/src/factory.rs:675     fn unmap_buffer_raw(&mut self, _map: RawMapping) {
                                         ^
src/backend/metal/src/factory.rs:675:5: 677:6 help: run `rustc --explain E0407` to see a detailed explanation
src/backend/metal/src/factory.rs:690:34: 690:74 error: type name `core::mapping::RW` is undefined or not in scope [E0412]
src/backend/metal/src/factory.rs:690                               -> core::mapping::RW<T, Resources, Factory> {
                                                                      ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
src/backend/metal/src/factory.rs:690:34: 690:74 help: run `rustc --explain E0412` to see a detailed explanation
src/backend/metal/src/factory.rs:690:34: 690:74 help: no candidates by the name of `RW` found in your project; maybe you misspelled the name or forgot to import an external crate?
src/backend/metal/src/command.rs:382:6: 382:39 error: trait `command::CommandBuffer` is not in scope [E0405]
src/backend/metal/src/command.rs:382 impl command::CommandBuffer<Resources> for CommandBuffer {
                                          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
src/backend/metal/src/command.rs:382:6: 382:39 help: run `rustc --explain E0405` to see a detailed explanation
src/backend/metal/src/command.rs:382:6: 382:39 help: no candidates by the name of `CommandBuffer` found in your project; maybe you misspelled the name or forgot to import an external crate?
src/backend/metal/src/map.rs:448:74: 448:87 error: unresolved name `NoMultiSample` [E0425]
src/backend/metal/src/map.rs:448         MTLTextureType::D2            => shade::TextureType::D2(NoArray, NoMultiSample),
                                                                                                          ^~~~~~~~~~~~~
src/backend/metal/src/map.rs:448:74: 448:87 help: run `rustc --explain E0425` to see a detailed explanation
src/backend/metal/src/map.rs:449:72: 449:85 error: unresolved name `NoMultiSample` [E0425]
src/backend/metal/src/map.rs:449         MTLTextureType::D2Array       => shade::TextureType::D2(Array, NoMultiSample),
                                                                                                        ^~~~~~~~~~~~~
src/backend/metal/src/map.rs:449:72: 449:85 help: run `rustc --explain E0425` to see a detailed explanation
src/backend/metal/src/map.rs:450:74: 450:85 error: unresolved name `MultiSample` [E0425]
src/backend/metal/src/map.rs:450         MTLTextureType::D2Multisample => shade::TextureType::D2(NoArray, MultiSample),
                                                                                                          ^~~~~~~~~~~
src/backend/metal/src/map.rs:450:74: 450:85 help: run `rustc --explain E0425` to see a detailed explanation
src/backend/metal/src/map.rs:460:22: 460:44 error: unresolved name `factory::RENDER_TARGET` [E0425]
src/backend/metal/src/map.rs:460     if bind.contains(factory::RENDER_TARGET) || bind.contains(factory::DEPTH_STENCIL) {
                                                      ^~~~~~~~~~~~~~~~~~~~~~
src/backend/metal/src/map.rs:460:22: 460:44 help: run `rustc --explain E0425` to see a detailed explanation
src/backend/metal/src/map.rs:460:63: 460:85 error: unresolved name `factory::DEPTH_STENCIL` [E0425]
src/backend/metal/src/map.rs:460     if bind.contains(factory::RENDER_TARGET) || bind.contains(factory::DEPTH_STENCIL) {
                                                                                               ^~~~~~~~~~~~~~~~~~~~~~
src/backend/metal/src/map.rs:460:63: 460:85 help: run `rustc --explain E0425` to see a detailed explanation
src/backend/metal/src/map.rs:464:22: 464:46 error: unresolved name `factory::SHADER_RESOURCE` [E0425]
src/backend/metal/src/map.rs:464     if bind.contains(factory::SHADER_RESOURCE) {
                                                      ^~~~~~~~~~~~~~~~~~~~~~~~
src/backend/metal/src/map.rs:464:22: 464:46 help: run `rustc --explain E0425` to see a detailed explanation
src/backend/metal/src/map.rs:468:22: 468:47 error: unresolved name `factory::UNORDERED_ACCESS` [E0425]
src/backend/metal/src/map.rs:468     if bind.contains(factory::UNORDERED_ACCESS) {
                                                      ^~~~~~~~~~~~~~~~~~~~~~~~~
src/backend/metal/src/map.rs:468:22: 468:47 help: run `rustc --explain E0425` to see a detailed explanation
src/backend/metal/src/map.rs:475:27: 475:36 error: type name `MapAccess` is undefined or not in scope [E0412]
src/backend/metal/src/map.rs:475 pub fn map_access(access: MapAccess) -> MTLResourceOptions {
                                                           ^~~~~~~~~
src/backend/metal/src/map.rs:475:27: 475:36 help: run `rustc --explain E0412` to see a detailed explanation
src/backend/metal/src/map.rs:475:27: 475:36 help: no candidates by the name of `MapAccess` found in your project; maybe you misspelled the name or forgot to import an external crate?
src/backend/metal/src/map.rs:477:9: 477:28 error: failed to resolve. Use of undeclared type or module `MapAccess` [E0433]
src/backend/metal/src/map.rs:477         MapAccess::Readable => MTLResourceCPUCacheModeDefaultCache,
                                         ^~~~~~~~~~~~~~~~~~~
src/backend/metal/src/map.rs:477:9: 477:28 help: run `rustc --explain E0433` to see a detailed explanation
src/backend/metal/src/map.rs:478:9: 478:28 error: failed to resolve. Use of undeclared type or module `MapAccess` [E0433]
src/backend/metal/src/map.rs:478         MapAccess::Writable => MTLResourceCPUCacheModeWriteCombined,
                                         ^~~~~~~~~~~~~~~~~~~
src/backend/metal/src/map.rs:478:9: 478:28 help: run `rustc --explain E0433` to see a detailed explanation
src/backend/metal/src/map.rs:479:9: 479:22 error: failed to resolve. Use of undeclared type or module `MapAccess` [E0433]
src/backend/metal/src/map.rs:479         MapAccess::RW => MTLResourceCPUCacheModeDefaultCache,
                                         ^~~~~~~~~~~~~
src/backend/metal/src/map.rs:479:9: 479:22 help: run `rustc --explain E0433` to see a detailed explanation
src/backend/metal/src/lib.rs:277:77: 277:92 error: `tex::Descriptor` does not name a structure [E0422]
src/backend/metal/src/lib.rs:277     let color_tex = device.share.handles.borrow_mut().make_texture(raw_tex, tex::Descriptor {
                                                                                                             ^~~~~~~~~~~~~~~
src/backend/metal/src/lib.rs:277:77: 277:92 help: run `rustc --explain E0422` to see a detailed explanation
error: cannot continue compilation due to previous error
Build failed, waiting for other jobs to finish...
error: Could not compile `gfx_device_metal`.
````

**Comment by Bastacyclop** [comment]
I made a PR that didn't correctly update the Metal backend (I have no way of testing it), so this is expected.
It should be mostly easy-to-fix errors (paths, names, signatures ...).
I don't know when this will be fixed but I think @fkaa is currently working on the Metal backend.

**Comment by kvark** [comment]
We should either fix Metal quickly or remove it from Cargo config until it becomes properly CI-tested.
@jamiltron you can try removing all metal-related stuff manually for now (from `Cargo.toml`) to get going while we are working on a solution. Sorry for the trouble!

**Comment by jamiltron** [comment]
No problem! I'll give it a try later today. Thanks!

**Comment by jamiltron** [comment]
There are numerous references to metal in src/lib.rs as well as src/shade.rs, but commenting out all of those references allows me to run the examples.

**Comment by fkaa** [comment]
Fixed in #1049
